feat: Bcrypt algorithm support

seremenko-wish opened the pull request:
## Related issue
#1137 @aeneasr 
<!--
Please link the GitHub issue this pull request resolves in the format of `#1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@aeneasr`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers either in the Github Discusssions in this repository or
join the [ORY Chat](https://www.ory.sh/chat).
-->

## Proposed changes
- New `hashers.algorithm` option in config to control password hashing algorithm. By default, the value is `argon2`. In order to use the Bcrypt hashing algorithm, `bcrypt` value should be used. Also, added a setting to control the cost for Bcrypt algorithm (`hashers.bcrypt.cost`) with default value `12`.
- Bcrypt algorithm truncates all input strings to 72 bytes, so a corresponding check was added
- Tests for Bcrypt hasher

## Checklist
- [X] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [X] I have read the [security policy](../security/policy).
- [X] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [X] I have added tests that prove my fix is effective or that my feature
      works.
- [X] I have added or changed [the documentation](docs/docs).

## Further comments
- Change of `hashers.algorithm`should be considered a breaking change since passwords hashed by Argon2 won't be properly validated after this change. In order to handle this case properly, hashers should be separated from classes that check hashes. Should we create a separate issue for this?
- Returning an error when a user tries to use very long password when the Bcrypt algorithm is active, seems like a user-friendly behavior. I would like to get feedback about this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ory/kratos?pullRequest=1169) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ory/kratos?pullRequest=1169) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ory/kratos?pullRequest=1169) it.</sub>

aeneasr reviewed (changes requested):
Thank you for working on this, it looks good! One thing we need to take care of however is that we can't change the algorithm if we have users signed up already, as those users would no longer be able to log in.

This means that we need a password upgrade here: https://github.com/ory/kratos/blob/master/selfservice/strategy/password/login.go#L149

The code for generating the password entry can be found at https://github.com/ory/kratos/blob/master/selfservice/strategy/password/registration.go#L177-L196

It would probably also make sense to add a section about bcrypt here: https://github.com/ory/kratos/blob/master/docs/docs/concepts/credentials/username-email-password.mdx

And lastly, it's actually fine by me if we use BCrypt per default, as the kratos-typical workloads are web based, meaning that Argon2 might underperform.

zepatrik commented on driver/config/config.go:
```diff
@@ -732,3 +748,14 @@ func (p *Config) PasswordPolicyConfig() *PasswordPolicy {
+	switch configValue {
+	case "bcrypt":
+		return configValue
+	case "argon2":
```
Switches in go break by default :wink: 
```suggestion
	case "argon2":
		fallthrough
```

seremenko-wish commented:
Last commit fixes https://github.com/ory/kratos/issues/1175

seremenko-wish commented:
> Thank you for working on this, it looks good! One thing we need to take care of however is that we can't change the algorithm if we have users signed up already, as those users would no longer be able to log in.

I separated the hash comparator from hashers, so changing of hashing algorithm won't prevent users from logging in.

> This means that we need a password upgrade here: https://github.com/ory/kratos/blob/master/selfservice/strategy/password/login.go#L149
fixed

> The code for generating the password entry can be found at https://github.com/ory/kratos/blob/master/selfservice/strategy/password/registration.go#L177-L196
this code does not need any changes as it uses the hasher specified in the configuration

> It would probably also make sense to add a section about bcrypt here: https://github.com/ory/kratos/blob/master/docs/docs/concepts/credentials/username-email-password.mdx
added

> And lastly, it's actually fine by me if we use BCrypt per default, as the kratos-typical workloads are web based, meaning that Argon2 might underperform.
For backwards compatibility, I left Argon2id as the default hasher. We can create a separate task to make that switch in the future.

zepatrik commented on hash/hasher_bcrypt.go:
```diff
@@ -0,0 +1,43 @@
+	// so if password is longer than 72 bytes, function returns an error
+	// See https://en.wikipedia.org/wiki/Bcrypt#User_input
+	if len(password) > 72 {
+		return errors.New("passwords are limited to a maximum length of 72 characters")
```
This error will be passed down to the UI I guess. Please add it to the messages to allow translations and mapping to other text: https://github.com/ory/kratos/blob/1940a679eb6b695e22cb939fb0d8d85cebb82a1e/text/message_registration.go

I guess you would then in the flow check for this error you return here similar to https://github.com/ory/kratos/blob/4bda1f09472bd65608cafe0d5d44287487da6f70/selfservice/flow/registration/error.go#L92

But maybe @aeneasr can give better pointers.

aeneasr replied:
Right, it would be an error like this one (actually we can just add it do that file for now): https://github.com/ory/kratos/blob/master/schema/errors.go#L59-L70

zepatrik reviewed (changes requested):
Nice, this looks really good :+1:

seremenko-wish commented on driver/config/.schema/config.schema.json:
```diff
@@ -1294,6 +1294,12 @@
+          "description": "One of the values: argon2, bcrypt",
+          "type": "string",
+          "default": "argon2"
+        },
```
I would set the minimum to 12 as well. There is no official recommendation, and I found a few posts in StackOverflow that point out that the minimum cost in 2021 for web-based password hashing shouldn't be less than 12. See this post for example https://stackoverflow.com/a/61304956

aeneasr commented on hash/hash_comparator.go:
```diff
@@ -0,0 +1,107 @@
+		return nil, nil, nil, err
+	}
+
+	salt, err = base64.RawStdEncoding.Strict().DecodeString(parts[4])
```
Will using `Strict()` here cause issues with existing hashes?

seremenko-wish replied:
No, all correct existing hashes will be ok. Strict() just does not allow to decode invalid base64 encoded strings.

aeneasr replied:
Great, thank you!

aeneasr reviewed (commented):
Awesome, just two tiny things but this is almost good to ship!

aeneasr reviewed (approved):
Awesome, thank you! 🎉 Your contribution makes Ory better :)

aeneasr commented:
The last thing is getting tests green - so running `make format` will resolve the one issue. The e2e tests have been failing for a while now on master so we need for someone to fix those first. I will probably not be able to do that this week :/

aeneasr commented:
I have resolved the e2e tests so this should work again - rerunning tests now.

seremenko-wish commented:
ok. all checks are green now!

aeneasr commented:
Awesome, thank you so much for your hard work!

harnash commented on hash/hash_comparator.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func isArgon2idHash(hash []byte) bool {
+	res, _ := regexp.Match("^\\$argon2id\\$", hash)
```
Nitpicking: wouldn't `strings.HasPrefix()` be sufficient in this case?

seremenko-wish replied:
Agree, that could be a good simplification. Not sure if it is worth doing it now. I used this approach as it will allow adding more algorithms in the future and user regexes for identifying them.

harnash replied:
Nah. I think is not worth separate PR right now. When/if we tackle with #605 then that would be a good time to refactor this. 

Anyway @seremenko-wish thanks a lot for this PR, this will get us closer to get all critical features implemented. Is the #1175 closable now since hash comparison is fixed in this PR?

seremenko-wish replied:
closed #1175

this definitely opens gates for more algorithm integrations. Looks like PBKDF1/2 integration is missing, but that would be easy to add into existing hashing framework.

aeneasr replied:
Yup! Not sure if we want to support other algorithms as main hasher but definitely support them when importing users/hashes! This PR lays the foundation for doing that!

seremenko-wish commented:
@aeneasr would you want to add this feature to the next version release?

aeneasr commented:
Yes, this is planned for 0.6

I am currently working on pushing out 0.6 this month!

aeneasr commented:
Looks like OWASP is also not quite clear themselves as to which algorithm to support as they have recently updated their docs to encourage use of Argon2id whenever possible: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pre-hashing-passwords

seremenko-wish commented:
Yeah. That's fine. I saw that they recommend some argon2 settings. It would be interesting to measure performance with those settings.

btw, they encourage not to use any pre-hashing functions with bcrypt https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pre-hashing-passwords

aeneasr commented:
Yes, thank you! I also saw that in a talk someone mentioned in another issue ;)
